internal/tracker/downstream_test.go

```diff
@@ -0,0 +1,109 @@
+package tracker
+
+import (
+	"reflect"
+	"sync"
+	"testing"
+)
+
+func TestDownstreamConnsCounts(t *testing.T) {
+	downstream1 := "downstream1"
+	downstream2 := "downstream2"
+
+	tests := []struct {
+		name           string
+		op             func(*DownstreamConns)
+		expectedCounts map[string]uint32
+	}{
+		{
+			name: "record a new downstream connection if under maximum",
+			op: func(tracker *DownstreamConns) {
+				tracker.TryBeginConnection(downstream1, 10)
+			},
+			expectedCounts: map[string]uint32{
+				downstream1: 1,
+			},
+		},
+		{
+			name: "record a connection ending",
+			op: func(tracker *DownstreamConns) {
+				tracker.TryBeginConnection(downstream1, 10)
+				tracker.EndConnection(downstream1)
+			},
+			expectedCounts: map[string]uint32{
+				downstream1: 0,
+			},
+		},
+		{
+			name: "don't record connections which would extend beyond maximum",
+			op: func(tracker *DownstreamConns) {
+				tracker.TryBeginConnection(downstream1, 10)
```

Comment:
This function returns a boolean. Any reason why it is not checked?

Reply:
Yes. I could test this more in depth, but I was willing to make an assumption that returning true will always accompany an additional connection and false will always accompany no additional connection. The connection counts which are checked at the end of the test cover the same case. 

This probably wouldn't meet my bar for productions code, especially for such a critical piece of code. I choose to not meet that standard because of this direction in the challenge document, and my time constraints.

> Add a couple of high quality tests that cover happy and unhappy scenarios.
>
> Do not try to achieve full test coverage. This will take too long. Take two key components, e.g. authentication/authorization layer and networking and implement one or two test cases that demonstrate your approach to testing.